sample_waveforms.py

```diff
@@ -506,3 +506,165 @@ def sigma(ell):
+        reencoded_version_hist = [
+            [entry[0].encode("ascii", "ignore"), entry[1].encode("ascii", "ignore")] for entry in version_hist
+        ]
+        dset = h5file.create_dataset(
+            "VersionHist.ver",
+            (len(version_hist), 2),
+            maxshape=(None, 2),
+            data=reencoded_version_hist,
+            dtype=h5py.special_dtype(vlen=bytes),
+        )
+        dset.attrs.create("Version", len(version_hist), shape=(1,), dtype=np.uint64)
+
+        # Generate the asymptotic waveform
+        h0 = scri.sample_waveforms.fake_precessing_waveform(
+            t_0=t_0, t_1=t_1, dt=dt, ell_max=ell_max, precession_opening_angle=precession_opening_angle, **kwargs
+        )
+        n_times = int(h0.t.shape[0] + r_max / dt)
+        t_1 += r_max
+        all_times = np.linspace(0, t_1, n_times)
+
+        # Set auxiliary datasetsss
+        for i in range(len(groups)):
+            # Set coordinate radius dataset
+            coord_radius = np.vstack((all_times, [coord_radii[i]] * n_times)).T
+            dset = h5file.create_dataset(groups[i] + "/CoordRadius.dat", data=coord_radius)
+            dset.attrs.create("Legend", ["time", "CoordRadius"])
+
+            # Set areal radius dataset
+            areal_radius = coord_radius
+            areal_radius[:, 1] += avg_areal_radius_diff
+            dset = h5file.create_dataset(groups[i] + "/ArealRadius.dat", data=areal_radius)
+            dset.attrs.create("Legend", ["time", "ArealRadius"])
+
+            # Set initial ADM energy dataset
+            seg_start_times = h0.t[:: int(n_times / 20)]
+            adm_energy_dset = np.vstack((seg_start_times, [initial_adm_energy] * len(seg_start_times))).T
+            dset = h5file.create_dataset(groups[i] + "/InitialAdmEnergy.dat", data=adm_energy_dset)
+            dset.attrs.create("Legend", ["time", "InitialAdmEnergy"])
+
+            # Set average lapse dataset
+            avg_lapse_dset = np.vstack((all_times, [avg_lapse] * n_times)).T
+            dset = h5file.create_dataset(groups[i] + "/AverageLapse.dat", data=avg_lapse_dset)
+            dset.attrs.create("Legend", ["time", "AverageLapse"])
+
+            # Set finite radius data
+            R = areal_radius[0,1]
+            tortoise_coord = R + 2 * initial_adm_energy * np.log(R / (2 * initial_adm_energy) - 1)
+            # Compute the approximate time in SpEC simulation coordinates
+            simulation_time = (h0.t + tortoise_coord) * np.sqrt(1 - 2 * initial_adm_energy / R) / avg_lapse
+            for l in range(2, ell_max + 1):
+                for m in range(-l, l + 1):
+                    index = sf.LM_index(l, m, 2)
+                    new_data = h0.data[:, index]
+                    for n in range(1, n_subleading + 1):
+                        new_data += amp * h0.abs[:, index] * R ** -n * np.exp((1j * n * 50 * np.pi / h0.n_times) * h0.t)
```

Comment:
I'm sure the difference is teensy here, so there's no need to change it, but FYI: the reason I put `h0.t` and `h0.abs[:, index]` at the ends of their respective expressions is because python performs operations left-to-right.  So if you put all the floats at the beginning, you're doing as many (fast) float operations as possible, and minimizing the number of (slow) array operations.  I've gotten significant speedups from that in some cases.  Just something to keep in mind when dealing with larger arrays in core functions.

Reply:
:scream: How did I not know this?! I'll throw this in with the change you've requested below, it takes two seconds.